tfhe/src/integer/server_key/radix_parallel/vector_find.rs

```diff
@@ -0,0 +1,1282 @@
+    ///
+    /// # Notes
+    ///
+    /// - clear values in the slice must be unique (otherwise use
+    ///   [Self::smart_first_index_in_clears_parallelized])
+    /// - If the encrypted value is not in the clear slice, the returned index is 0
+    pub fn smart_index_in_clears_parallelized<T, Clear>(
+        &self,
+        ct: &mut T,
+        clears: &[Clear],
+    ) -> (RadixCiphertext, BooleanBlock)
+    where
+        T: IntegerRadixCiphertext,
+        Clear: DecomposableInto<u64> + CastInto<usize>,
+    {
+        if !ct.block_carries_are_empty() {
+            self.full_propagate_parallelized(ct);
+        }
+
+        self.unchecked_index_in_clears_parallelized(ct, clears)
+    }
+
+    /// Returns the encrypted index of the encrypted `value` in the clear slice
+    /// also returns an encrypted boolean that is `true` if the encrypted value was found.
+    ///
+    /// # Notes
+    ///
+    /// - clear values in the slice must be unique (otherwise use
+    ///   [Self::index_in_clears_parallelized])
+    /// - If the encrypted value is not in the clear slice, the returned index is 0
+    pub fn index_in_clears_parallelized<T, Clear>(
+        &self,
+        ct: &T,
+        clears: &[Clear],
+    ) -> (RadixCiphertext, BooleanBlock)
+    where
+        T: IntegerRadixCiphertext,
+        Clear: DecomposableInto<u64> + CastInto<usize>,
+    {
+        let mut tmp_ct;
+        let ct = if ct.block_carries_are_empty() {
+            ct
+        } else {
+            tmp_ct = ct.clone();
+            self.full_propagate_parallelized(&mut tmp_ct);
+            &tmp_ct
+        };
+
+        self.unchecked_index_in_clears_parallelized(ct, clears)
+    }
+
+    /// Returns the encrypted index of the _first_ occurrence of encrypted `value` in the clear
+    /// slice also, it returns an encrypted boolean that is `true` if the encrypted value was
+    /// found.
+    ///
+    /// # Notes
+    ///
+    /// - If the encrypted value is not in the clear slice, the returned index is 0
+    pub fn unchecked_first_index_in_clears_parallelized<T, Clear>(
+        &self,
+        ct: &T,
+        clears: &[Clear],
+    ) -> (RadixCiphertext, BooleanBlock)
+    where
+        T: IntegerRadixCiphertext,
+        Clear: DecomposableInto<u64> + CastInto<usize> + Hash,
+    {
+        if clears.is_empty() {
+            return (
+                self.create_trivial_zero_radix(ct.blocks().len()),
+                self.create_trivial_boolean_block(false),
+            );
+        }
+        let unique_clears = clears
+            .iter()
+            .copied()
+            .enumerate()
+            .unique_by(|&(_, value)| value)
+            .collect::<Vec<_>>();
+        let selectors = self.compute_equality_selectors(
+            ct,
+            unique_clears.par_iter().copied().map(|(_, value)| value),
+        );
+        let selectors2 = selectors.iter().cloned().map(|x| x.0).collect::<Vec<_>>();
```

Comment:
Yes they are useful for 2 reasons:
1) is that 1 function expects boolean block the other does not:
  technically  `is_at_least_one_comparisons_block_true` could be changed to require BooleanBlocks

2) Most important reason: `is_at_least_one_comparisons_block_true` takes a `Vec<Ciphertext>` because it reuses
the memory of the vec and ciphertexts while reducing it to a single output, meaning even if we changed its input to `Vec<BooleanBlock>` we would need a clone